### PR TITLE
Display a message when a page created by an application can't be properly displayed because of a missing license

### PR DESCRIPTION
**Pull request by arcilli**
[Issue #40](https://github.com/xwikisas/application-diagram/issues/40): Diagram is not displayed

As @mflorea suggested, in the situation when user has no license for a page, we should display his raw content. Also, we will display a message containing the applications without license.

### CODE REVIEW

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+      <async_context/>
+    </property>
+    <property>
+      <async_enabled>0</async_enabled>
+    </property>
+    <property>
+      <content>{{velocity}}
+#set ($sheetsMissingLicense = $services.licensing.licensor.getSheetsMissingLicense($doc.documentReference, $xcontext.action))
+#if ($services.sheet.getSheets($doc, $xcontext.action).isEmpty() &amp;&amp; $sheetsMissingLicense.size() &gt; 0)
+  #set($message = '')
```
Comment (acotiuga) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: Use `#set (` (a blank space), to be consistent

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+  #foreach ($sheet in $sheetsMissingLicense)
+    #foreach ($extension in $services.extension.xar.getInstalledExtensions($sheet))
+      #set ($message = $message+" "+$extension.getName())
```
Comment (acotiuga) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: I don't see how you will get the representation of a list, without a separator. For readability you could use some spaces. Also `$extension.name` is the same as `$extension.getName()`, but shorter.
You can use an array to store the extension names and then to join them with a separator in the translation key instead of concatenating strings.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+     #end
```
Comment (acotiuga) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: Wrong indentation here.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+  #end
+  {{html clean="false"}}
+  &lt;div class="hidden" id="licenseCheckingExtensionPoint"&gt;$services.localization.render("licensor.wrongDisplayedContent", [$message])&lt;/div&gt;
```
Comment (acotiuga) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: You can improve the indentation here to something like:
```
<div class="hidden" id="licenseCheckingExtensionPoint">
  $services.localization.render("licensor.wrongDisplayedContent", [$message])
</div>
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    </property>
+    <property>
+      <async_context/>
+    </property>
+    <property>
+      <async_enabled>0</async_enabled>
+    </property>
+    <property>
+      <content>{{velocity}}
+#set ($sheetsMissingLicense = $services.licensing.licensor.getSheetsMissingLicense($doc.documentReference, $xcontext.action))
```
Comment (acotiuga) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: you need to call `getUnlicensedSheets()` here (after renaming).

File: application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java [hunk 6]
```diff
@@ -237,4 +249,22 @@ public Long getUserCount()
...
+    public List<DocumentReference> getUnlicensedSheets(DocumentReference documentReference, String action)
+    {
+        try {
+            DocumentModelBridge documentModelBridge = documentAccessBridge.getDocument(documentReference);
+            List<DocumentReference> sheetsUsedByDocument = sheetManager.getSheets(documentModelBridge, action);
+            return sheetsUsedByDocument.stream()
+                    .filter(sheet -> !this.hasLicensureForEntity(sheet))
+                    .collect(Collectors.toList());
+        } catch (Exception e) {
+            return null;
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java]: We could also log an error here so it's easier to debug the problems that may occur.

File: application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java [hunk 7]
```diff
@@ -237,4 +249,22 @@ public Long getUserCount()
             return null;
         }
     }
+
+    /**
+     * @param documentReference is the document for which is searching for associated sheets with missing license
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java]:
```suggestion
     * @param documentReference the document for which to search for associated sheets with missing license
```

File: application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java [hunk 8]
```diff
@@ -237,4 +249,22 @@ public Long getUserCount()
...
+     * @param action represents the action made
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java]:
```suggestion
     * @param action represents the action the sheets have to match
```

File: application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java [hunk 9]
```diff
@@ -237,4 +249,22 @@ public Long getUserCount()
...
+     * @return a list of DocumentReference (sheets) that are missing license for a specific documentReference and action
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/script/LicensorScriptService.java]:
```suggestion
     * @return the list of sheets associated with the specified document and action and that are missing a license
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/Translations.xml [hunk 10]
```diff
@@ -93,7 +93,8 @@ licensor.updateLicenses.failedToRetrieve=Failed to retrieve the license updates.
 licensor.updateLicenses.noUpdates=The licenses are already up-to-date.
 licensor.updateLicenses.applied=Applied {0} out of {1} license updates.
 licensor.badCSRFToken=Bad CSRF token
-licensor.addLicense.submit=Add License</content>
+licensor.addLicense.submit=Add License
+licensor.wrongDisplayedContent=The content of this page is not correctly displayed because the following extensions does not have a valid license: {0}.</content>
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/Translations.xml]:
```suggestion
licensor.wrongDisplayedContent=This page is not correctly displayed because the following extensions don''t have a valid license: {0}.</content>
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+        <classType>com.xpn.xwiki.objects.classes.StaticListClass</classType>
+      </use>
+    </class>
+    <property>
+      <cache>long</cache>
+    </property>
+    <property>
+      <code>require(['jquery'], function($) {
+  $(document).ready(function() {
+    $('#licenseCheckingExtensionPoint').detach().prependTo('#xwikicontent');
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: You don't need to ``detach``. This is done automatically by ``prependTo``.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+    $('#licenseCheckingExtensionPoint').removeClass('hidden');
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: I would have called ``removeClass`` on the previous line so that I don't lookup the ``licenseCheckingExtensionPoint`` element twice.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+-->
+
+<xwikidoc version="1.3" reference="Licenses.Code.UIExtensionLicenseChecking" locale="">
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: By looking at https://github.com/xwikisas/application-licensing/tree/master/application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code I think the naming convention used so far is to have the UIX suffix for UIExtension pages. So in order to be consistent I would go for ``LicenseCheckingUIX``. **But**, I see that we already have a page ``MissingLicensesUIX`` which uses the same extension point (after header) so maybe we should extend that page instead of creating another one. WDYT?

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+      <async_context/>
+    </property>
+    <property>
+      <async_enabled>0</async_enabled>
+    </property>
+    <property>
+      <content>{{velocity}}
+#set ($unlicensedSheets = $services.licensing.licensor.getUnlicensedSheets($doc.documentReference, $xcontext.action))
+#if ($services.sheet.getSheets($doc, $xcontext.action).isEmpty() &amp;&amp; $unlicensedSheets.size() &gt; 0)
+  #set ($extensions = [])
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: I would use a set here to make sure we don't get duplicates.

```suggestion
  #set ($extensions = $collectionstool.set)
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+  #foreach ($sheet in $unlicensedSheets)
+    #foreach ($extension in $services.extension.xar.getInstalledExtensions($sheet))
+      #set ($discard = $extensions.add($extension.name))
+    #end
+  #end
+  {{html clean="false"}}
+  &lt;div class="hidden" id="licenseCheckingExtensionPoint"&gt;
+    $services.localization.render("licensor.wrongDisplayedContent", [$stringtool.join($extensions, ',')])
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: I believe you also need a space after comma.

```suggestion
    $services.localization.render("licensor.wrongDisplayedContent", [$stringtool.join($extensions, ', ')])
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+  &lt;/div&gt;
+  {{/html}}
+  #set($discard = $xwiki.jsx.use('Licenses.Code.UIExtensionLicenseChecking', {'minify':false}))
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: Keep ``{'minify':false}`` only while debugging. Remove it before committing.

```suggestion
  #set($discard = $xwiki.jsx.use('Licenses.Code.UIExtensionLicenseChecking'))
```

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml (hunk 13 above)
Comment (arcilli) [reply application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/UIExtensionLicenseChecking.xml]: Since both use the same EP, probably is a good idea to merge the code in `MissingLicensesUIX`.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml [hunk 17]
```diff
@@ -120,6 +223,22 @@
...
+#set ($unlicensedSheets = $services.licensing.licensor.getUnlicensedSheets($doc.documentReference, $xcontext.action))
+#if ($services.sheet.getSheets($doc, $xcontext.action).isEmpty() &amp;&amp; $unlicensedSheets.size() &gt; 0)
+  #set ($extensions = $collectionstool.set)
+  #foreach ($sheet in $unlicensedSheets)
+    #foreach ($extension in $services.extension.xar.getInstalledExtensions($sheet))
+      #set ($discard = $extensions.add($extension.name))
+    #end
+  #end
+  {{html clean="false"}}
+  &lt;div class="hidden" id="licenseCheckingExtensionPoint"&gt;
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml]: I would add a comment explaining why we hide the message and then move it with JavaScript (since it's not obvious and it's easy to forget in 1 month): because the "After content header" UIX (``org.xwiki.platform.template.content.header.after``) is available only since XWiki 10.4RC1 and we need to support XWiki 8.4+

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml [hunk 18]
```diff
@@ -120,6 +223,25 @@
...
+  #set ($extensions = $collectionstool.set)
+  #foreach ($sheet in $unlicensedSheets)
+    #foreach ($extension in $services.extension.xar.getInstalledExtensions($sheet))
+      #set ($discard = $extensions.add($extension.name))
+    #end
+  #end
+  {{html clean="false"}}
+  ## Create a div that will display a warning and the applications that don't have a license.
+  ## Because the "After container header" UIX (org.xwiki.platform.template.content.header.after) is available since
+  ## XWiki 10.4RC1 and we need to support XWiki 8.4+, the div will be created and with hidden class. After that, it will bemoved with JS.
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml]: Line too long (>121 chars) and there are some typos: "will be created and with" and "bemoved". You should review your changes before committing.

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml [hunk 19]
```diff
@@ -40,6 +40,109 @@
...
+  $(document).ready(function() {
+    $('#licenseCheckingExtensionPoint').removeClass('hidden').prependTo('#xwikicontent');
+  });
+});</code>
+    </property>
+    <property>
+      <name/>
+    </property>
+    <property>
+      <parse/>
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml]: You should set to No, since you don't have any Velocity code in your JSX (at the moment at least).

File: application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml [hunk 20]
```diff
@@ -40,6 +40,109 @@
...
+    </property>
+    <property>
+      <use/>
```
Comment (mflorea) [inline application-licensing-licensor/application-licensing-licensor-ui/src/main/resources/Licenses/Code/MissingLicensesUIX.xml]: You should set to "On demand only".
